src/libsystemd-network/arp-util.c

```diff
@@ -48,12 +49,12 @@ int arp_network_bind_raw_socket(int ifindex, be32_t address, const struct ether_
                 BPF_JUMP(BPF_JMP + BPF_JEQ + BPF_K, ARPOP_REPLY, 1, 0),                        /* protocol == reply ? */
                 BPF_STMT(BPF_RET + BPF_K, 0),                                                  /* ignore */
                 /* Sender Hardware Address must be different from our own */
-                BPF_STMT(BPF_LD + BPF_IMM, htobe32(*((uint32_t *) eth_mac))),                  /* A <- 4 bytes of client's MAC */
+                BPF_STMT(BPF_LD + BPF_IMM, unaligned_read_be32(&eth_mac->ether_addr_octet[0])),/* A <- 4 bytes of client's MAC */
```

Comment:
i am a bit confused about the old code here, it uses "htobe32()" instead of "be32toh()", and that looks wrong...

unaligned_read_be32() of course is a wrapper around b32toh(), hence your patch actually fixes not just one but two bugs. 

That said, I am not a BPF guru... Can you clarify what endianess BPF actually uses? I assume native, and hence your patch is right, but maybe it doesn't?

Reply:
The old code couldn't possibly be right because mac addresses are always big endian.

Reply:
I'm pretty sure it's native. I think neither old nor proposed new code is entirely correct, it should just do unaligned read without byte swapping.

Reply:
@topimiettinen 
This is only the *first* 32-bits of a 48-bit address.

Reply:
For the first MAC of 01:02:03:04:05:06 and second of aa:bb:cc:dd:ee:ff, we'd want a constant of 0x04030201 on little endian for the first 32 bits, so it can be compared to the value in BPF register loaded with non-swapped access (0xddccbbaa), no?